src/Stratis.Bitcoin.Features.Consensus/Rules/ProvenHeaderRules/ProvenHeaderCoinstakeRule.cs

```diff
@@ -0,0 +1,281 @@
+        /// </exception>
+        private void CheckHeaderAndCoinstakeTimes(ProvenBlockHeader header)
+        {
+            uint coinstakeTime = header.Coinstake.Time;
+            uint headerTime = header.Time;
+
+            // Check if times are equal and coinstake tx time is divisible by 16.
+            if ((headerTime == coinstakeTime) && ((coinstakeTime & PosConsensusOptions.StakeTimestampMask) == 0))
+                return;
+
+            this.Logger.LogTrace("(-)[BAD_TIME]");
+            ConsensusErrors.StakeTimeViolation.Throw();
+        }
+
+        /// <summary>
+        /// Checks the coinstake age requirement.
+        /// </summary>
+        /// <param name="chainedHeader">The chained header.</param>
+        /// <param name="unspentOutputs">The unspent outputs.</param>
+        /// <exception cref="ConsensusException">
+        /// Throws exception with error <see cref="ConsensusErrors.InvalidStakeDepth" /> if check fails.
+        /// </exception>
+        private void CheckCoinstakeAgeRequirement(ChainedHeader chainedHeader, UnspentOutputs unspentOutputs)
+        {
+            ChainedHeader prevChainedHeader = chainedHeader.Previous;
+
+            var options = (PosConsensusOptions)this.PosParent.Network.Consensus.Options;
+            int targetDepth = options.GetStakeMinConfirmations(chainedHeader.Height, this.PosParent.Network) - 1;
+
+            if(this.stakeValidator.IsConfirmedInNPrevBlocks(unspentOutputs, prevChainedHeader, targetDepth))
+                return;
+
+            this.Logger.LogTrace("(-)[BAD_STAKE_DEPTH]");
+            ConsensusErrors.InvalidStakeDepth.Throw();
+        }
+
+        /// <see cref="IStakeValidator.VerifySignature"/>
+        /// <exception cref="ConsensusException">
+        /// Throws exception with error <see cref="ConsensusErrors.CoinstakeVerifySignatureFailed" /> if check fails.
+        /// </exception>
+        private void CheckSignature(ProvenBlockHeader header, UnspentOutputs unspentOutputs)
+        {
+            if (this.stakeValidator.VerifySignature(unspentOutputs, header.Coinstake, 0, ScriptVerify.None))
+                return;
+            
+            this.Logger.LogTrace("(-)[BAD_SIGNATURE]");
+            ConsensusErrors.CoinstakeVerifySignatureFailed.Throw();
+        }
+
+        /// <see cref="IStakeValidator.CheckStakeKernelHash"/>
+        /// <exception cref="ConsensusException">
+        /// Throws exception with error <see cref="ConsensusErrors.PrevStakeNull" /> if check fails.
+        /// </exception>
+        private void CheckStakeKernelHash(PosRuleContext context, UnspentOutputs stakingCoins, ProvenBlockHeader header, ChainedHeader chainedHeader)
+        {
+            OutPoint prevOut = this.GetPreviousOut(header);
+            uint transactionTime = header.Coinstake.Time;
+
+            ChainedHeader prevChainedHeader = chainedHeader.Previous;
+
+            BlockStake prevBlockStake = this.PosParent.StakeChain.Get(prevChainedHeader.HashBlock);
```

Comment:
This is going to be problematic, this is a header rule so prevChainedHeader.HashBlock may not exist in stake store, stake store will only have the stake info of the last connected block.

Reply:
Ok after discussion with @noescape00 we realised there is a design bug here.
https://github.com/stratisproject/StratisBitcoinFullNode/issues/2500
So we merge this PR as is and address the design bug in a separate issue.